src/group_theory/iwasawa.lean

```diff
@@ -0,0 +1,486 @@
+
+theorem is_maximal_iff {K: subgroup G} : K.is_maximal ↔
+  K ≠ ⊤ ∧ ∀ (H: subgroup G) g, K ≤ H → g ∉ K → g ∈ H → H = ⊤ :=
+begin
+  split,
+  { intro hK, split, exact is_maximal.ne_top hK,
+  intros H g hKH hgK hgH,
+  suffices : K < H,
+    apply (is_maximal_def.1 hK).2, assumption,
+  have zKH : K ≠ H, { rw ne.def, intro z, rw z at hgK, exact hgK hgH },
+  exact (ne.le_iff_lt zKH).mp hKH,},
+  { rintros ⟨ hG, hmax ⟩,
+  split, split, assumption,
+  introsI H hKH,
+  obtain ⟨ g, hgH, hgK ⟩ := (set.exists_of_ssubset hKH),
+  exact hmax H g (le_of_lt hKH) hgK hgH, },
+end
+
+theorem is_maximal.eq_of_le {K H: subgroup G}
+  (hK : K.is_maximal) (hH : H ≠ ⊤) (KH : K ≤ H) : K = H :=
+eq_iff_le_not_lt.2 ⟨KH, λ h, hH (hK.1.2 _ h)⟩
+
+end subgroup
+end Maximal_Subgroups
+
+section Basic_Lemmas
+/-- If G and H have multiplications *
+and φ : G → H is a surjective multiplicative map,
+and if G is commutative, then H is commutative.
+
+Since I only use it with groups,
+I should probably use function.surjective.comm_semigroup
+--/
+lemma surj_to_comm {G H : Type*} [has_mul G] [has_mul H] (φ: mul_hom G H) :
+   function.surjective φ → is_commutative G (*) → is_commutative H (*) :=
+begin
+  intros is_surj is_comm,
+  apply is_commutative.mk,
+  intros a b,
+    obtain ⟨a', ha'⟩ := is_surj a, obtain ⟨b', hb'⟩ := is_surj b,
+    rw ← ha', rw ← hb',
+    let z := ⇑φ, let z₂ := φ.to_fun, have : z = z₂  , by refl,
+    rw ← mul_hom.map_mul φ a' b',
+    rw ← mul_hom.map_mul φ b' a',
+    apply φ.congr_arg,
+    refine is_commutative.cases_on is_comm _, intro,
+    exact comm a' b',
+end
+
+/- No more used
+
+/-- f(f⁻¹(t)) = f.range ∩ t  -/
+lemma image_preimage_eq {α β : Type*} (f: α → β) (t: set β):
+  set.image f (set.preimage f t) = set.range f ∩ t :=
+begin
+    sorry,
+end
+
+lemma image_preimage_eq' {α β : Type*} (f: α → β) (t: set β):
+  set.image f (set.preimage f t) = set.range f ∩ t :=
+  by rw [← set.image_univ, ← set.image_inter_preimage f set.univ t, set.univ_inter]
+
+/-- If f is surjective, then f(f⁻¹(t)) = t -/
+--  is in mathlib :  set.image_preimage_eq
+lemma image_preimage_eq_of_surjective' {α β : Type*} (f : α → β) (t : set β) :
+  function.surjective f → set.image f (set.preimage f t) = t :=  λ hf,
+  by rw [image_preimage_eq f t, set.range_iff_surjective.2 hf,  set.univ_inter]
+ -/
+end Basic_Lemmas
+
+section Primitive_And_Transitive_Actions
+open_locale big_operators pointwise
+open function
+
+open set mul_action subgroup
+
+variables (G : Type*) [group G] (X : Type*) [mul_action G X]
+
+/-- Transitive actions -/
+structure is_transitive
+extends is_pretransitive G X : Prop :=
+(is_nonempty : nonempty X)
+
+/-
+lemma nonempty_of_transitive : is_transitive G X → nonempty X
+:= λ h, h.is_nonempty
+
+lemma pretransitive_of_transitive : is_transitive G X → is_pretransitive G X :=
+λ h, h.is_pretransitive
+-/
+
+variables (G X)
+structure is_primitive
+extends is_transitive G X : Prop :=
```

Comment:
Why not just use `is_pretransitive`? The only serious issue arises in the proof of `Iwasawa_mk`, but there you can deduce `nonempty X` from `mul_action.fixed_points N X ≠ ⊤`.
```suggestion
extends is_pretransitive G X : Prop :=
```

Reply:
The name of the structure should reflect its mathematical name.
As you notice, the Iwasawa lemma uses two properties of `is_primitive`, `is_transitive` and `has_maximal_stabilizers`, while the third one, as you indicate, could be derived from the other hypothesis. 
Does it make sense to define `is_preprimitive` which would just remove the `is_nonempty` field?

Reply:
Either `is_primitive` or `is_preprimitive` is fine with me.